src/plasma_manager.c

```diff
@@ -163,101 +376,284 @@ void read_object_chunk(event_loop *loop,
-  utstring_printf(ip_addr_port, "%s:%d", utstring_body(ip_addr), port);
-  client_connection *manager_conn;
-  HASH_FIND_STR(conn->manager_state->manager_connections,
-                utstring_body(ip_addr_port), manager_conn);
-
-  if (!manager_conn) {
-    /* If we don't already have a connection to this manager, start one. */
-    manager_conn = malloc(sizeof(client_connection));
-    manager_conn->fd = plasma_manager_connect(utstring_body(ip_addr), port);
-    manager_conn->manager_state = conn->manager_state;
-    manager_conn->transfer_queue = NULL;
-    manager_conn->cursor = 0;
-
-    manager_conn->ip_addr_port = strdup(utstring_body(ip_addr_port));
-    HASH_ADD_KEYPTR(hh, manager_conn->manager_state->manager_connections,
-                    manager_conn->ip_addr_port,
-                    strlen(manager_conn->ip_addr_port), manager_conn);
-  }
-  utstring_free(ip_addr_port);
+  client_connection *manager_conn =
+      get_manager_connection(conn->manager_state, utstring_body(ip_addr), port);
   utstring_free(ip_addr);
 
   if (manager_conn->transfer_queue == NULL) {
     /* If we already have a connection to this manager and its inactive,
      * (re)register it with the event loop again. */
     event_loop_add_file(loop, manager_conn->fd, EVENT_LOOP_WRITE,
-                        write_object_chunk, manager_conn);
+                        send_buffered_request, manager_conn);
   }
   /* Add this transfer request to this connection's transfer queue. */
   LL_APPEND(manager_conn->transfer_queue, buf);
 }
 
-void start_reading_data(event_loop *loop,
-                        int client_sock,
-                        object_id object_id,
-                        int64_t data_size,
-                        int64_t metadata_size,
-                        client_connection *conn) {
-  plasma_buffer *buf = malloc(sizeof(plasma_buffer));
+void process_data(event_loop *loop,
+                  int client_sock,
+                  object_id object_id,
+                  int64_t data_size,
+                  int64_t metadata_size,
+                  client_connection *conn) {
+  plasma_request_buffer *buf = malloc(sizeof(plasma_request_buffer));
   buf->object_id = object_id;
   buf->data_size = data_size;
   buf->metadata_size = metadata_size;
-  buf->writable = 1;
 
-  plasma_create(conn->manager_state->store_conn, object_id, data_size, NULL,
+  plasma_create(conn->manager_state->plasma_conn, object_id, data_size, NULL,
                 metadata_size, &(buf->data));
   LL_APPEND(conn->transfer_queue, buf);
   conn->cursor = 0;
 
   /* Switch to reading the data from this socket, instead of listening for
    * other requests. */
   event_loop_remove_file(loop, client_sock);
-  event_loop_add_file(loop, client_sock, EVENT_LOOP_READ, read_object_chunk,
+  event_loop_add_file(loop, client_sock, EVENT_LOOP_READ, process_data_chunk,
                       conn);
 }
 
+void request_transfer_from(client_connection *client_conn,
```

Comment:
how about request_transfer_from_next_manager (that is what it is doing, right?); it is a little long...